cpmonitor/views.py

```diff
@@ -506,6 +508,49 @@ def markdown_uploader_view(request):
     return HttpResponse(data, content_type="application/json")
 
 
+def mstr_view(request, municipality_key):
+    url = "https://www.marktstammdatenregister.de/MaStR/Einheit/EinheitJson/GetVerkleinerteOeffentlicheEinheitStromerzeugung"
+    params = {
+        "sort": "EinheitMeldeDatum-desc",
+        "page": 1,
+        "pageSize": 10000,
+        "filter": f"Energieträger~eq~'2495'~and~Gemeindeschlüssel~eq~'{municipality_key}'",
+    }
+    r = requests.get(url, params)
+    data = r.json()["Data"]
+
+    installed_by_year = dict()
+
+    for entry in data:
+        if not entry["InbetriebnahmeDatum"]:
+            continue
+
+        m = re.match("/Date\((-?\d*)\)/", entry["InbetriebnahmeDatum"])
+        install_date = datetime.fromtimestamp(int(m.group(1)) / 1000)
+
+        if not install_date.year in installed_by_year:
+            installed_by_year[install_date.year] = 0
+
+        installed_by_year[install_date.year] += entry["Bruttoleistung"]
+
+    START_YEAR = 2019
+    current_year = date.today().year
+    installed_accumulated = 0
+    years = []
+    installed = []
+    for year, installed_in_year in sorted(installed_by_year.items()):
```

Comment:
I would maybe move the year filter to the for loop above (it would keep the dictionary smaller).
And maybe we can add some unit test?

Reply:
The filtered years (i.e. years before START_YEAR) are still accumulated, so they cannot be simply filtered earlier. However, I changed the first loop so that accumulation of years up to START_YEAR happens immediately. This might be a big harder to read, though, so take a look, if you think it is more difficult to understand now I would revert to the previous version. After all, the dictionary is not really large (max. ~30 keys and integer values for each key)

Reply:
Actually right now I just meant to extract the data conversion part into a function and unit test that function (like test_status_proportions_should_have_the_right_sorting_order_when_all_statuses_are_present)